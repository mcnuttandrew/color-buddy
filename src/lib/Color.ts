import type { Color as ChromaColor } from "chroma-js";
import chroma from "chroma-js";
import ColorIO from "colorjs.io";

const hexCache = new Map<string, string>();
export class Color {
  name: string = "";
  channels: Record<string, number> = {};
  chromaBind: typeof chroma.lab = chroma.rgb;
  spaceName: keyof typeof colorDirectory = "rgb";

  toHex(): string {
    const str = this.toString();
    if (hexCache.has(str)) {
      return hexCache.get(str) as string;
    }
    const newHex = this.toColorIO().to("srgb").toString({ format: "hex" });
    // const oldHex = this.chromaBind(...this.toChannels()).hex();
    // return this.chromaBind(...this.toChannels()).hex();
    hexCache.set(str, newHex);
    return newHex;
  }
  toString(): string {
    const channelsString = Object.values(this.channels)
      .map((x) => x || 0)
      .map((x) => (x < 1e-7 ? 0 : x))
      .join(", ");
    return `${this.spaceName}(${channelsString})`;
  }
  getChannel(channel: keyof typeof this.channels): number {
    return this.channels[channel];
  }
  toChroma(): ChromaColor {
    return this.chromaBind(...this.toChannels());
  }
  toChannels(): [number, number, number] {
    return Object.values(this.channels) as [number, number, number];
  }
  setChannel(channel: keyof typeof this.channels, value: number) {
    this.channels[channel] = value;
  }
  inGamut(): boolean {
    // return new ColorIO(this.spaceName, this.toChannels()).inGamut();
    const x = this.toHex();
    const y = colorFromHex(x, "srgb").toHex();
    // if (x !== y) {
    //   console.log("x", x, "y", y);
    // }
    return x === y;
  }
  toColorIO(): ColorIO {
    return new ColorIO(this.toString());
  }

  fromString(colorString: string): Color {
    if (!colorString.startsWith(`${this.spaceName}(`)) {
      const channels = new ColorIO(colorString).to(this.spaceName).coords;
      return this.fromChannels(channels);
    }
    return this.fromChannels(stringToChannels(this.spaceName, colorString));
  }
  fromChannels(channels: [number, number, number]): Color {
    const newColor = new (this.constructor as typeof Color)();
    Object.keys(this.channels).forEach((channel, i) => {
      newColor.setChannel(channel, channels[i]);
    });
    return newColor;
  }
  luminance(): number {
    return chroma(this.toHex()).luminance();
  }
  deltaE(color: Color): number {
    return chroma.deltaE(this.toChroma(), color.toChroma());
  }
  symmetricDeltaE(color: Color): number {
    return 0.5 * (this.deltaE(color) + color.deltaE(this));
  }
  copy(): Color {
    return this.fromChannels(this.toChannels());
  }
}

const colorStringCache = new Map<string, [number, number, number]>();
export function stringToChannels(spaceName: string, str: string) {
  const key = `${spaceName}(${str})`;
  if (colorStringCache.has(key)) {
    return colorStringCache.get(key)!;
  }
  let channels = str
    .replace(`${spaceName}(`, "")
    .replace(")", "")
    .replace(/%/g, "")
    .replace(/,/g, " ")
    .trim()
    .split(" ")
    .filter((x) => x.length);
  const allNumbers = channels.every((x) => !isNaN(+x));

  if (!(channels.length === 3 && allNumbers)) {
    throw new Error(`Invalid color string: ${str}`);
  }
  const result = channels.map((x) => Number(x) * 1) as [number, number, number];
  colorStringCache.set(key, result);
  return result;
}

export class CIELAB extends Color {
  name = "CIELAB";
  channels = { L: 0, a: 0, b: 0 };
  chromaBind = chroma.lab;
  spaceName = "lab" as const;

  toString(): string {
    const [L, a, b] = Object.values(this.channels).map((x) => x || 0);
    // .map((x) => x.toPrecision(1));
    return `lab(${L}% ${a} ${b})`;
  }
}
export class HSV extends Color {
  name = "HSV";
  channels = { h: 0, s: 0, v: 0 };
  chromaBind = chroma.hsv;
  spaceName = "hsv" as const;
  toString(): string {
    const [h, s, v] = Object.values(this.channels);
    return `color(hsv ${h} ${s} ${v})`;
  }
}

export class RGB extends Color {
  name = "RGB";
  channels = { r: 0, g: 0, b: 0 };
  chromaBind = chroma.rgb;
  spaceName = "rgb" as const;
}

export class HSL extends Color {
  name = "HSL";
  channels = { h: 0, s: 0, l: 0 };
  chromaBind = chroma.hsl;
  spaceName = "hsl" as const;

  toString(): string {
    const [h, s, l] = Object.values(this.channels).map((x) => x || 0);
    return `hsl(${h} ${s}% ${l}%)`;
  }
}
export class LCH extends Color {
  name = "LCH";
  channels = { l: 0, c: 0, h: 0 };
  chromaBind = chroma.lch;
  spaceName = "lch" as const;
}

export class OKLAB extends Color {
  name = "OKLAB";
  channels = { l: 0, a: 0, b: 0 };
  chromaBind = chroma.oklab;
  spaceName = "oklab" as const;
}

export class OKLCH extends Color {
  name = "OKLCH";
  channels = { l: 0, c: 0, h: 0 };
  chromaBind = chroma.oklch;
  spaceName = "oklch" as const;
}

export class JZAZBZ extends Color {
  name = "JZAZBZ";
  channels = { jz: 0, az: 0, bz: 0 };
  spaceName = "jzazbz" as const;
  constructor() {
    super();
    this.chromaBind = (jz, az, bz) => {
      const [l, a, b] = new ColorIO(this.toString()).to("lab").coords;
      return chroma.lab(l, a, b);
    };
  }
  toString(): string {
    const [jz, az, bz] = Object.values(this.channels);
    return `color(jzazbz ${jz} ${az} ${bz})`;
  }
}

export function colorFromString(
  colorString: string,
  colorSpace: keyof typeof colorDirectory = "lab"
): Color {
  return new colorDirectory[colorSpace]().fromString(colorString);
}

const colorHexCache = new Map<string, Color>();
export function colorFromHex(
  hex: string,
  colorSpace: keyof typeof colorDirectory
): Color {
  if (colorHexCache.has(hex)) {
    return colorHexCache.get(hex)!.copy() as Color;
  }
  const color = new ColorIO(hex).to(colorSpace);
  const outColor = new colorDirectory[colorSpace]().fromChannels(color.coords);
  colorHexCache.set(hex, outColor);
  return outColor;
}

export function colorFromChannels(
  channels: [number, number, number],
  colorSpace: keyof typeof colorDirectory
): Color {
  return new colorDirectory[colorSpace]().fromChannels(channels);
}

export function toColorSpace(
  color: Color,
  colorSpace: keyof typeof colorDirectory
): Color {
  if (color.spaceName === colorSpace) {
    return color;
  }
  const channels = new ColorIO(color.toString()).to(colorSpace).coords;
  return new colorDirectory[colorSpace]().fromChannels(channels);
}

export const colorDirectory = {
  hsl: HSL,
  hsv: HSV,
  jzazbz: JZAZBZ,
  lab: CIELAB,
  lch: LCH,
  oklab: OKLAB,
  oklch: OKLCH,
  rgb: RGB,
  srgb: RGB,
};

type ColorSpace = keyof typeof colorDirectory | string;
const domains = {
  lab: { a: [-125, 125], b: [-125, 125], l: [0, 100] },
  oklab: { a: [-0.4, 0.4], b: [-0.4, 0.4], l: [0, 1] },
  rgb: { r: [0, 255], g: [0, 255], b: [0, 255] },
  // srgb: { r: [0, 255], g: [255, 0], b: [0, 255] },
  hsv: { h: [0, 360], s: [0, 1], v: [0, 1] },
  hsl: { h: [0, 360], s: [0, 1], l: [0, 1] },
  lch: { l: [0, 100], c: [0, 150], h: [0, 360] },
  oklch: { l: [0, 1], c: [0, 0.4], h: [0, 360] },
  jzazbz: { jz: [0, 1], az: [-0.5, 0.5], bz: [-0.5, 0.5] },
} as Record<ColorSpace, Record<string, [number, number]>>;
export const stepSize = {
  lab: [1, 1, 1],
  oklab: [0.01, 0.01, 0.01],
  rgb: [1, 1, 1],
  // srgb: [1, 1, 1],
  hsv: [1, 0.01, 0.01],
  hsl: [1, 0.01, 0.01],
  lch: [1, 1, 1],
  oklch: [0.01, 0.01, 1],
  jzazbz: [0.01, 0.01, 0.01],
} as Record<ColorSpace, [number, number, number]>;

const dimensionToChannel = {
  lab: { x: "a", y: "b", z: "l" },
  oklab: { x: "a", y: "b", z: "l" },
  rgb: { x: "g", y: "b", z: "r" },
  // srgb: { x: "g", y: "b", z: "r" },
  hsv: { x: "s", y: "v", z: "h" },
  hsl: { x: "s", y: "l", z: "h" },
  lch: { x: "c", y: "h", z: "l" },
  oklch: { x: "c", y: "h", z: "l" },
  jzazbz: { x: "az", y: "bz", z: "jz" },
} as Record<ColorSpace, Record<string, string>>;
const xyTitles = {
  lab: "CIELAB: a* b*",
  oklab: "OKLAB: a* b*",
  rgb: "RGB: Green Blue",
  // srgb: "sRGB: Green Blue",
  hsv: "HSV: Saturation Value",
  hsl: "HSL: Saturation Lightness",
  lch: "LCH: Chroma Hue",
  oklch: "OKLCH: Chroma Hue",
  jzazbz: "JzAzBz: Az Bz",
} as Record<ColorSpace, string>;
const zTitles = {
  lab: "CIELAB: L*",
  oklab: "OKLAB: L*",
  rgb: "RGB: Red",
  // srgb: "sRGB: Red",
  hsv: "HSV: Hue",
  hsl: "HSL: Hue",
  lch: "LCH: Lightness",
  oklch: "OKLCH: Lightness",
  jzazbz: "JzAzBz: Jz",
} as Record<ColorSpace, string>;
const titleMap = {
  lab: "CIELAB",
  oklab: "OKLAB",
  rgb: "RGB",
  // srgb: "sRGB",
  hsv: "HSV",
  hsl: "HSL",
  lch: "LCH",
  oklch: "OKLCH",
  jzazbz: "JzAzBz",
} as Record<ColorSpace, string>;

export const colorPickerConfig = Object.fromEntries(
  (Object.keys(domains) as ColorSpace[]).map((name) => {
    return [
      name,
      {
        title: titleMap[name],
        xyTitle: xyTitles[name],
        zTitle: zTitles[name],
        xDomain: domains[name][dimensionToChannel[name].x],
        yDomain: domains[name][dimensionToChannel[name].y],
        zDomain: domains[name][dimensionToChannel[name].z],
        xChannel: dimensionToChannel[name].x,
        yChannel: dimensionToChannel[name].y,
        zChannel: dimensionToChannel[name].z,
        xStep: stepSize[name][1],
        yStep: stepSize[name][2],
        zStep: stepSize[name][0],
      },
    ];
  })
);
