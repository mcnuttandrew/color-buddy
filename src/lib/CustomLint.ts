import { ColorLint } from "./ColorLint";
import type { PalType, Affect, Context, Palette } from "../types";
import {
  LLEval,
  prettyPrintLL,
  permutativeBlame,
} from "./lint-language/lint-language";

import * as Json from "jsonc-parser";
export interface CustomLint {
  affectTypes?: Affect[];
  contextTypes?: Context[];
  blameMode: "pair" | "single" | "none";
  description: string;
  failMessage: string;
  group: string;
  id: string;
  level: "error" | "warning";
  name: string;
  program: string;
  subscribedFix?: string;
  taskTypes: PalType[];
  expectedPassingTests: Palette[];
  expectedFailingTests: Palette[];
}

export function CreateCustomLint(props: CustomLint) {
  let natProg = "";
  try {
    natProg = prettyPrintLL(Json.parse(props.program));
  } catch (e) {}
  return class CustomLint extends ColorLint<number[] | number[][], any> {
    name = props.name;
    taskTypes = props.taskTypes;
    affectTypes = props.affectTypes || [];
    contextTypes = props.contextTypes || [];
    level = props.level;
    group = props.group;
    description = props.description;
    isCustom = props.id;
    blameMode = props.blameMode;
    subscribedFix = props.subscribedFix || "none";
    naturalLanguageProgram = natProg;

    _runCheck(options: any) {
      const prog = Json.parse(props.program);
      const { blame, result } = LLEval(prog, this.palette, {
        debugCompare: false,
        ...options,
      });
      if (result) return { passCheck: true, data: blame };
      let newBlame: number[] | number[][] = [];
      if (this.blameMode !== "none") {
        newBlame = permutativeBlame(prog, this.palette, this.blameMode);
      }

      return { passCheck: result, data: newBlame };
    }

    buildMessage() {
      let blame = "";
      if (this.blameMode === "pair") {
        blame = (this.checkData as number[][])
          .map((x) =>
            x.map((x) => this.palette.colors[x].color.toHex()).join(" and ")
          )
          .join(", ");
      } else {
        blame = (this.checkData as number[])
          .map((x) => this.palette.colors[x].color.toHex())
          .join(", ");
      }

      return props.failMessage.replaceAll("{{blame}}", blame);
    }
  };
}
