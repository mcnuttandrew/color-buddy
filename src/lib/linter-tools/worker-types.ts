export type WorkerCommand =
  | { type: "load-lints"; content: string; id: string }
  | { type: "run-lint"; content: string; id: string }
  | { type: "run-lint-no-message"; content: string; id: string }
  | { type: "monte-carlo-fix"; content: string; id: string };
