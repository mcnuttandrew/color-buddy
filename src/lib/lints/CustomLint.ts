import type { LintProgram } from "../lint-language/lint-type";
import { ColorLint } from "./ColorLint";
import type { TaskType } from "./ColorLint";
import { LLEval, prettyPrintLL } from "../lint-language/lint-language";

export interface CustomLint {
  program: LintProgram;
  name: string;
  taskTypes: TaskType[];
  level: "error" | "warning";
  group: string;
  description: string;
  failMessage: string;
}

export function CreateCustomLint(props: CustomLint) {
  return class CustomLint extends ColorLint<number[], any> {
    name = props.name;
    taskTypes = props.taskTypes;
    level = props.level;
    group = props.group;
    description = props.description;
    hasHeuristicFix = false;
    isCustom = true;

    _runCheck() {
      const { blame, result } = LLEval(props.program, this.palette, {
        debugCompare: false,
      });
      return { passCheck: result, data: blame };
    }

    buildMessage() {
      const blame = this.checkData
        .map((x) => this.palette.colors[x].toHex())
        .join(", ");
      return props.failMessage.replace("{{blame}}", blame);
    }
  };
}
