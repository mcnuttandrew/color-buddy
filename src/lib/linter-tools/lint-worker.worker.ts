import * as idb from "idb-keyval";
import { runLintChecks } from "../linter";
import type { CustomLint } from "../CustomLint";
import type { Palette } from "../../types";
import { Color } from "../Color";
import type { LintResult } from "../ColorLint";

type Command =
  | { type: "load-lints"; content: ""; id: string }
  | { type: "run-lint"; content: ""; id: string };

const cache: Record<string, Color> = {};
const getColor = (hex: string, space: string): Color => {
  const key = `${hex}-${space}`;
  if (cache[key]) {
    return cache[key];
  }
  const result = Color.colorFromString(hex, space as any);
  cache[key] = result;
  return result;
};

const hydratePal = (pal: string): Palette => {
  const parsed = JSON.parse(pal);
  return {
    background: getColor(parsed.background, parsed.colorSpace),
    colors: parsed.colors.map((x: string) => getColor(x, parsed.colorSpace)),
    type: parsed.type,
    colorSpace: parsed.colorSpace,
    name: parsed.name,
    evalConfig: parsed.evalConfig,
    intendedAffects: parsed.intendedAffects,
    intendedContexts: parsed.intendedContexts,
  };
};

let lintStore: CustomLint[] = [];
let storeLoaded = false;
const storeName = "color-pal-lints";
let simpleLintCache = new Map<string, any>();
async function dispatch(cmd: Command) {
  switch (cmd.type) {
    case "load-lints":
      idb.get(storeName).then((x) => {
        lintStore = x.lints;
        storeLoaded = true;
        // gotta dump the cache when new lints are loaded
        simpleLintCache = new Map<string, any>();
        return "";
      });
      return "";
    case "run-lint":
      if (simpleLintCache.has(cmd.content)) {
        return simpleLintCache.get(cmd.content);
      }
      const pal = hydratePal(cmd.content);
      // if store not loaded, wait
      while (!storeLoaded) {
        await new Promise((resolve) => setTimeout(resolve, 100));
      }

      const result: LintResult[] = runLintChecks(pal, lintStore).map((x) => {
        return {
          name: x.name,
          passes: x.passes,
          message: x.message,
          level: x.level,
          group: x.group,
          description: x.description,
          isCustom: x.isCustom,
          taskTypes: x.taskTypes as any,
          subscribedFix: x.subscribedFix,
        };
      });
      simpleLintCache.set(cmd.content, result);
      return result;
    default:
      return "no-op";
  }
}

self.onmessage = async (event: MessageEvent<Command>) => {
  const result = await dispatch(event.data);
  self.postMessage({ id: event.data.id, content: result });
};

export {}; // this is to make typescript happy
