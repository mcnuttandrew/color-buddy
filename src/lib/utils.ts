import chroma from "chroma-js";
import { Color, colorFromString, colorFromChannels } from "./Color";
import type { PalType } from "../stores/color-store";
export const insert = (arr: Color[], newItem: Color, index?: number) => {
  if (index === undefined) {
    return [...arr, newItem];
  }
  return [...arr.slice(0, index), newItem, ...arr.slice(index)];
};

export const replaceVal = (arr: Color[], newItem: Color, index: number) => {
  return [...arr.slice(0, index), newItem, ...arr.slice(index + 1)];
};

export const randChan = () => Math.floor(Math.random() * 255);
export const randColor = () =>
  colorFromString(`rgb(${randChan()},${randChan()},${randChan()})`, "lab");

// seeded random
export const seededPick = (seedInit: number) => {
  var seed = seedInit;
  function random() {
    var x = Math.sin(seed++) * 10000;
    return x - Math.floor(x);
  }
  return (arr: any[]) => arr[Math.floor(random() * arr.length)];
};
export const pick = (arr: any[]) => arr[Math.floor(Math.random() * arr.length)];

export function avgColors(
  colors: Color[],
  colorSpace: "rgb" | "hsl" | "lab" = "rgb"
): Color {
  if (colors.length === 0) {
    return colorFromString("#000000", "lab");
  }

  type ColorChannels = [number, number, number];
  const sum = (a: ColorChannels, b: ColorChannels) =>
    a.map((x, i) => x + b[i]) as ColorChannels;
  const sumColor = colors.reduce((acc, x) => sum(acc, x.toChannels()), [
    0, 0, 0,
  ] as ColorChannels);
  const avgColor = sumColor.map((x) => x / colors.length) as ColorChannels;
  if (avgColor.some((x) => isNaN(x))) {
    return colorFromString("#000000", colorSpace);
  }
  return colorFromChannels(avgColor, colorSpace);
}

export function opposingColor(color: Color): Color {
  const c = color.toChroma().hsl();
  const channels = c.map((x, i) => (i === 0 ? (x + 180) % 360 : x));
  const chromaColor = chroma.hsl(channels[0], channels[1], channels[2]);
  return colorFromChannels(chromaColor.lab(), "lab");
}

export function deDup(arr: Color[]): Color[] {
  const seen = new Set();
  return arr.filter((item) => {
    const k = item.toHex();
    return seen.has(k) ? false : seen.add(k);
  });
}

export function draggable(node: any) {
  let x: number;
  let y: number;
  function handleMousedown(event: {
    type: string;
    touches: any[];
    clientX: any;
    clientY: any;
  }) {
    if (event.type === "touchstart") {
      event = event.touches[0];
    }
    x = event.clientX;
    y = event.clientY;
    node.dispatchEvent(new CustomEvent("dragstart", { detail: { x, y } }));
    // @ts-ignore
    window.addEventListener("mousemove", handleMousemove);
    // @ts-ignore
    window.addEventListener("mouseup", handleMouseup);
    // @ts-ignore
    window.addEventListener("touchmove", handleMousemove);
    // @ts-ignore
    window.addEventListener("touchend", handleMouseup);
  }
  function handleMousemove(event: {
    type: string;
    changedTouches: any[];
    clientX: number;
    clientY: number;
  }) {
    if (event.type === "touchmove") {
      event = event.changedTouches[0];
    }
    const dx = event.clientX - x;
    const dy = event.clientY - y;
    x = event.clientX;
    y = event.clientY;
    node.dispatchEvent(
      new CustomEvent("dragmove", {
        detail: { x, y, dx, dy },
      })
    );
  }
  function handleMouseup(event: { clientX: any; clientY: any }) {
    x = event.clientX;
    y = event.clientY;
    node.dispatchEvent(
      new CustomEvent("dragend", {
        detail: { x, y },
      })
    );
    // @ts-ignore
    window.removeEventListener("mousemove", handleMousemove);
    // @ts-ignore
    window.removeEventListener("mouseup", handleMouseup);
    // @ts-ignore
    window.removeEventListener("touchmove", handleMousemove);
    // @ts-ignore
    window.removeEventListener("touchend", handleMouseup);
  }
  node.addEventListener("mousedown", handleMousedown);
  node.addEventListener("touchstart", handleMousedown);
  return {
    destroy() {
      node.removeEventListener("mousedown", handleMousedown);
      node.removeEventListener("touchstart", handleMousedown);
    },
  };
}

const extent = (arr: number[]) => [Math.min(...arr), Math.max(...arr)];
export function makeExtents(arr: number[][]) {
  return {
    x: extent(arr.map((x) => x[1])),
    y: extent(arr.map((x) => x[2])),
    z: extent(arr.map((x) => x[0])),
  };
}

export const clamp = (n: number, min: number, max: number) =>
  Math.min(Math.max(n, min), max);

export const toggleElement = (arr: number[], el: number) => {
  const arrSet = new Set(arr);
  return arrSet.has(el) ? arr.filter((x) => x !== el) : [...arr, el];
};

export const swap = (arr: any[], i: number, j: number) => {
  const newArr = [...arr];
  const temp = newArr[i];
  newArr[i] = newArr[j];
  newArr[j] = temp;
  return newArr;
};

const colorBrewerTypeMap = {
  diverging: [
    "Spectral",
    "RdYlGn",
    "RdBu",
    "PiYG",
    "PRGn",
    "RdYlBu",
    "BrBG",
    "RdGy",
    "PuOr",
  ],

  categorical: [
    "Set2",
    "Accent",
    "Set1",
    "Set3",
    "Dark2",
    "Paired",
    "Pastel2",
    "Pastel1",
  ],

  sequential: [
    "OrRd",
    "PuBu",
    "BuPu",
    "Oranges",
    "BuGn",
    "YlOrBr",
    "YlGn",
    "Reds",
    "RdPu",
    "Greens",
    "YlGnBu",
    "Purples",
    "GnBu",
    "Greys",
    "YlOrRd",
    "PuRd",
    "Blues",
    "PuBuGn",
    "viridis",
  ],
};
export const colorBrewerMapToType = Object.entries(colorBrewerTypeMap).reduce(
  (acc, [type, maps]) => {
    maps.forEach((map) => {
      acc[map.toLowerCase()] = type.toLowerCase();
    });
    return acc;
  },
  {} as Record<string, PalType>
);
