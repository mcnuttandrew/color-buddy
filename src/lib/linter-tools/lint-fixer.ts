import type { Palette } from "../../types";
import type { LintResult } from "../ColorLint";
import { suggestFix, suggestMonteFix } from "../api-calls";
import { Color } from "../Color";
import { wrapInBlankSemantics } from "../utils";

type SuggestFix = (
  pal: Palette,
  lint: LintResult,
  engine: string
) => Promise<Palette[]>;

function parsePalette(colors: string[], colorSpace: Palette["colorSpace"]) {
  return colors.map((x) =>
    wrapInBlankSemantics(Color.colorFromHex(x.replace("##", "#"), colorSpace))
  );
}

export const suggestLintAIFix: SuggestFix = async (palette, lint, engine) => {
  const colorSpace = palette.colorSpace;
  const msg = `${lint.message}\n\nFailed: ${lint.naturalLanguageProgram}`;
  return suggestFix(palette, msg, engine as any).then((x) => {
    if (x.length === 0) {
      throw new Error("No suggestions");
    }
    return x.map((el) => {
      try {
        return { ...palette, colors: parsePalette(el.colors, colorSpace) };
      } catch (e) {
        console.log(e);
        return palette;
      }
    });
  });
};

export const suggestLintMonteFix: SuggestFix = async (palette, lint) => {
  const colorSpace = palette.colorSpace;
  // todo abstract out this logic
  return suggestMonteFix(palette, [lint.id!]).then((newPal) => {
    if (newPal.length === 0) {
      throw new Error("No suggestions");
    }
    try {
      return [{ ...palette, colors: parsePalette(newPal, colorSpace) }];
    } catch (e) {
      console.log(e);
      return [palette];
    }
  });
};

export const suggestMultiLintMonteFix = async (
  palette: Palette,
  lints: LintResult[]
) => {
  const colorSpace = palette.colorSpace;
  return suggestMonteFix(
    palette,
    lints.map((x) => x.id!)
  ).then((newPal) => {
    if (newPal.length === 0) {
      throw new Error("No suggestions");
    }
    try {
      return [{ ...palette, colors: parsePalette(newPal, colorSpace) }];
    } catch (e) {
      console.log(e);
      return [palette];
    }
  });
};

export type LintFixer = (pal: Palette, lint: LintResult) => Promise<Palette[]>;
import { fixBackgroundDifferentiability } from "../lints/background-contrast";
import { fixColorNameDiscriminability } from "../lints/name-discrim";
import { fixDivergingOrder } from "../lints/diverging-order";
import { fixGamut } from "../lints/in-gamut";
import { fixMaxColors } from "../lints/max-colors";
import { fixSequentialOrder } from "../lints/sequential-order";
import {
  fixLightnessDistribution,
  fixHueDistribution,
} from "../lints/even-distribution";
const fixDirectory: Record<string, LintFixer> = {
  fixBackgroundDifferentiability,
  fixColorNameDiscriminability,
  fixDivergingOrder,
  fixGamut,
  fixHueDistribution,
  fixLightnessDistribution,
  fixMaxColors,
  fixSequentialOrder,
};

export async function suggestLintFix(
  palette: Palette,
  lint: LintResult,
  _engine?: string
): Promise<Palette[]> {
  if (fixDirectory[lint.subscribedFix]) {
    return fixDirectory[lint.subscribedFix](palette, lint);
  }
  console.log("check failed", lint, fixDirectory);
  return [];
}
