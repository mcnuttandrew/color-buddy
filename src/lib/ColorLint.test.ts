import { expect, test } from "vitest";
import fs from "fs/promises";

import { Color } from "./Color";

import { CreateCustomLint } from "./CustomLint";
import { suggestLintFix } from "./linter-tools/lint-fixer";
import { makePalFromString } from "./utils";

import type { CustomLint } from "./CustomLint";

// Lints
import AvoidExtremes from "./lints/avoid-extremes";
import BackgroundContrast from "./lints/background-contrast";
import CatOrderSimilarity from "./lints/cat-order-similarity";
import ColorBlindness from "./lints/color-blindness";
import ColorNameDiscriminability, { getName } from "./lints/name-discrim";
import EvenDistribution from "./lints/even-distribution";
import Fair from "./lints/fair";
import Gamut from "./lints/in-gamut";
import MaxColors from "./lints/max-colors";
import MuthGuidelines from "./lints/muth-guidelines";
import MutuallyDistinct from "./lints/mutually-distinct";
import SequentialOrder from "./lints/sequential-order";
import SizeDiscrims from "./lints/size-discrim";
import UglyColors from "./lints/ugly-colors";

const unique = <T>(arr: T[]): T[] => [...new Set(arr)];

function autoTest(lint: CustomLint) {
  lint.expectedPassingTests.forEach((pal) => {
    const newLint = CreateCustomLint(lint);
    const exampleLint = new newLint(pal).run();
    expect(exampleLint.passes).toBe(true);
    expect(exampleLint.message).toMatchSnapshot();
  });
  lint.expectedFailingTests.forEach((pal) => {
    const newLint = CreateCustomLint(lint);
    const exampleLint = new newLint(pal).run();
    expect(exampleLint.passes).toBe(false);
    expect(exampleLint.message).toMatchSnapshot();
  });
  expect(lint.expectedFailingTests.length).toBeGreaterThan(0);
  expect(lint.expectedPassingTests.length).toBeGreaterThan(0);
}

test("ColorLint - AvoidExtremes", () => {
  autoTest(AvoidExtremes);
});

test("ColorLint - MutuallyDistinct", () => {
  autoTest(MutuallyDistinct);
});

test("ColorLint - MaxColors", () => {
  autoTest(MaxColors);
});

test("ColorLint - UglyColors", () => {
  autoTest(UglyColors);
});

test("ColorLint - EvenDistribution", () => {
  autoTest(EvenDistribution);
});

test("ColorLint - Fair Nominal", () => {
  autoTest(Fair[0]);
});

test("ColorLint - Fair Sequential", () => {
  autoTest(Fair[1]);
});

test("ColorLint - SequentialOrder", () => {
  autoTest(SequentialOrder);
});

test("ColorLint - CatOrderSimilarity", () => {
  autoTest(CatOrderSimilarity);
});

test("ColorLint - ColorNameDiscriminability", async () => {
  autoTest(ColorNameDiscriminability);

  const examplePal = makePalFromString(["#5260d1", "#005ebe"]);
  const lint = CreateCustomLint(ColorNameDiscriminability);
  const exampleLint = new lint(examplePal).run();
  expect(exampleLint.passes).toBe(false);
  expect(exampleLint.message).toBe(
    "The following pairs of colors have the same name: #5260d1 and #005ebe"
  );
  const fix = await suggestLintFix(examplePal, exampleLint);
  const oldColorNames = unique<string>(
    examplePal.colors.map((x) => getName(x))
  );
  expect(oldColorNames.length).toBe(1);
  const colorNames = unique<string>(fix[0].colors.map((x) => getName(x)));
  expect(colorNames.length).toBe(2);
});

test("ColorLint - SizeDiscrim (Thin)", () => {
  autoTest(SizeDiscrims[0]);
});

test("ColorLint - SizeDiscrim (Medium)", () => {
  autoTest(SizeDiscrims[1]);
});

test("ColorLint - SizeDiscrim (Wide)", () => {
  autoTest(SizeDiscrims[2]);
});

test("ColorLint - Gamut", async () => {
  autoTest(Gamut);
  const examplePal = makePalFromString(["lab(50.625% -91.737 -88.303)"]);
  const newLint = CreateCustomLint(Gamut);
  const exampleLint = new newLint(examplePal).run();
  expect(exampleLint.passes).toBe(false);
  expect(exampleLint.message).toMatchSnapshot();

  const fix = await suggestLintFix(examplePal, exampleLint);
  expect(fix[0].colors.map((x) => x.toString())).toStrictEqual([
    "lab(51.296% -23.327 -25.373)",
  ]);
});

test("ColorLint - ColorBlind: Deuteranopia", async () => {
  autoTest(ColorBlindness[0]);
});
test("ColorLint - ColorBlind: Protanopia", async () => {
  autoTest(ColorBlindness[1]);
});
test("ColorLint - ColorBlind: Tritanopia", async () => {
  autoTest(ColorBlindness[2]);
});
test("ColorLint - ColorBlind: Grayscale", async () => {
  autoTest(ColorBlindness[3]);
});

const ughWhat = ["#00ffff", "#00faff", "#00e4ff", "#fdfdfc", "#00ffff"];
test("ColorLint - Background Contrast", async () => {
  const examplePal = makePalFromString(ughWhat);
  const BackgroundContrastLint = CreateCustomLint(BackgroundContrast);
  const exampleLint = new BackgroundContrastLint(examplePal).run();
  expect(exampleLint.passes).toBe(false);
  expect(exampleLint.message).toBe(
    "These colors (#fdfdfc) do not have a sufficient contrast ratio with the background and may be hard to discriminate in some contexts."
  );
  const fix = await suggestLintFix(examplePal, exampleLint).then((x) => x[0]);
  expect(fix.colors.map((x) => x.toHex())).toMatchSnapshot();

  examplePal.background = Color.colorFromHex("#00e4ff", "lab");
  const exampleLint2 = new BackgroundContrastLint(examplePal).run();
  expect(exampleLint2.passes).toBe(false);
  expect(exampleLint2.message).toBe(
    "These colors (#00e4ff) do not have a sufficient contrast ratio with the background and may be hard to discriminate in some contexts."
  );
  const fix2 = await suggestLintFix(examplePal, exampleLint2).then((x) => x[0]);
  expect(fix2.colors.map((x) => x.toHex())).toMatchSnapshot();
  autoTest(BackgroundContrast);
});

test("ColorLint - MuthGuidelines (1) Background desaturation sufficient", () => {
  autoTest(MuthGuidelines[0]);
});

test("ColorLint - MuthGuidelines (2) Avoid Tetradic Palettes", () => {
  autoTest(MuthGuidelines[1]);
});

test("ColorLint - MuthGuidelines (3) Prefer yellowish or blueish greens", () => {
  autoTest(MuthGuidelines[2]);
});

test("ColorLint - MuthGuidelines (4) Avoid too much contrast with the background", () => {
  autoTest(MuthGuidelines[3]);
});
