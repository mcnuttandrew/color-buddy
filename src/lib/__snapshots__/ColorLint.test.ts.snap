// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorLint - BackgroundDifferentiability 1`] = `
[
  "#0ff",
  "#00faff",
  "#00e4ff",
  "#777776",
  "#0ff",
]
`;

exports[`ColorLint - BackgroundDifferentiability 2`] = `
[
  "#007070",
  "#006f72",
  "#006e7e",
  "#fdfdfc",
  "#007070",
]
`;

exports[`ColorLint - ColorBlind 1`] = `"This palette is not colorblind friendly for deuteranopia color blindness (ie can't see green). The following pairs are undifferentiable: (#0078b4 and #8c69bc), (#ff7e0e and #c4bc27), (#3d9f2f and #da2827), (#e179c1 and #00becf)"`;

exports[`ColorLint - ColorBlind 2`] = `"This palette is not colorblind friendly for protanopia color blindness (ie can't see red). The following pairs are undifferentiable: (#0078b4 and #8c69bc), (#ff7e0e and #3d9f2f), (#da2827 and #8e564b)"`;

exports[`ColorLint - ColorBlind 3`] = `"This palette is not colorblind friendly for tritanopia color blindness (ie can't see blue). The following pairs are undifferentiable: (#ff7e0e and #e179c1), (#8c69bc and #7f7f7f)"`;
